### ISSUE TITLE
Runway Plots

### ISSUE THREAD

**Issue by vaakesan-SMH** [VectorInstitute/cyclops#535]
### Is your feature request related to a problem? Please describe.
Most hard metrics for classifier model performance (F1, precision, recall, sensitivity, specificity, NPV) require a pre-specified classification threshold cutoff (usually 0.5). Although for some use cases, it is useful to see how the model performs, and what tradeoffs you can make, if you'd like to prioritize for example, true positives, or minimize for example, false positives.

### Describe the solution you'd like
One option is a runway plot, like the one shown below:

![Runway-plot-of-diagnostic-accuracy-measures-of-the-final-model](https://github.com/VectorInstitute/cyclops/assets/104445289/ebc15b29-834f-4fa9-b3a6-4b5d96bd6635)

### Describe alternatives you've considered
We can currently do this in the model transparency report by developing a `plotly` object and logging it to the report with `log_plotly_figure()`.

### Additional context
Below is a working function that will create a runway plot in plotly as required:

```
import plotly.graph_objects as go
import plotly.subplots as sp
import numpy as np
from typing import List
from sklearn.metrics import roc_curve, precision_recall_curve, confusion_matrix

def runway_plot(true_labels: List[int], pred_probs: List[float]) -> go.Figure:
    """
    Plot diagnostic performance metrics with an additional histogram of predicted probabilities.
    The plot uses Plotly with a clean aesthetic. Gridlines are kept, but background color is removed.
    Y-axis ticks and labels are shown. The legend is added at the bottom.
    Tooltips show values with 3 decimal places. X-axis labels are only shown on the bottom subplot.
    The histogram's bin size is reduced and it has no borders.

    Args:
    - true_labels (List[int]): True binary class labels (0 or 1).
    - pred_probs (List[float]): Predicted probabilities for the positive class (1).

    Returns:
    - A Plotly figure containing the diagnostic performance plots and histogram.

    Example:
    ```
    # Generate synthetic data for demonstration
    true_labels = np.random.binomial(1, 0.5, 1000)
    pred_probs = np.random.uniform(0, 1, 1000)

    # Generate and show the modified faceted plot
    faceted_fig = plot_diagnostic_performance_with_histogram(true_labels, pred_probs)
    faceted_fig.show()
    ```
    """
    
    # ROC curve components
    fpr, tpr, _ = roc_curve(true_labels, pred_probs)
    # Precision-Recall curve components
    precision, recall, _ = precision_recall_curve(true_labels, pred_probs)
    
    # Thresholds for PPV and NPV
    thresholds = np.linspace(0, 1, 100)
    ppv = np.zeros_like(thresholds)
    npv = np.zeros_like(thresholds)
    
    # Calculate PPV and NPV for each threshold
    for i, threshold in enumerate(thresholds):
        # Binarize predictions based on threshold
        binarized_predictions = pred_probs >= threshold
        tn, fp, fn, tp = confusion_matrix(true_labels, binarized_predictions).ravel()
        # Calculate PPV and NPV
        ppv[i] = tp / (tp + fp) if (tp + fp) != 0 else 0
        npv[i] = tn / (tn + fn) if (tn + fn) != 0 else 0

    # Define hover template to show three decimal places
    hover_template = 'Threshold: %{x:.3f}<br>Metric Value: %{y:.3f}<extra></extra>'
        
    # Create a subplot for each metric
    fig = sp.make_subplots(rows=5, cols=1, shared_xaxes=True, vertical_spacing=0.02)

    # Sensitivity plot (True Positive Rate)
    fig.add_trace(go.Scatter(x=thresholds, y=tpr, mode='lines', name='Sensitivity', hovertemplate=hover_template), row=1, col=1)
    
    # Specificity plot (1 - False Positive Rate)
    fig.add_trace(go.Scatter(x=thresholds, y=1-fpr, mode='lines', name='1 - Specificity', hovertemplate=hover_template), row=2, col=1)
    
    # PPV plot (Positive Predictive Value)
    fig.add_trace(go.Scatter(x=thresholds, y=ppv, mode='lines', name='PPV', hovertemplate=hover_template), row=3, col=1)
    
    # NPV plot (Negative Predictive Value)
    fig.add_trace(go.Scatter(x=thresholds, y=npv, mode='lines', name='NPV', hovertemplate=hover_template), row=4, col=1)
    
    # Add histogram of predicted probabilities
    fig.add_trace(go.Histogram(x=pred_probs, nbinsx=80, name='Predicted Probabilities'), row=5, col=1)

    # Update layout
    fig.update_layout(
        height=1000, 
        width=700, 
        title_text="Diagnostic Performance Metrics by Thresholds",
        legend=dict(orientation="h", yanchor="bottom", y=-0.2, xanchor="center", x=0.5)
    )

    # Remove subplot titles
    for i in fig['layout']['annotations']:
        i['text'] = ''

    # Remove the plot background color, keep gridlines, show y-axis ticks and labels
    fig.update_xaxes(showgrid=True)
    fig.update_yaxes(showgrid=True, showticklabels=True)

    # Only show the x-axis line and labels on the bottommost plot
    fig.update_xaxes(showline=True, linewidth=1, linecolor='black', mirror=True)
    fig.update_xaxes(showticklabels=True, row=4, col=1)
    fig.update_yaxes(showline=True, linewidth=1, linecolor='black', mirror=True)
    
    fig.update_xaxes(showline=False, row=5, col=1, showticklabels=False)
    fig.update_yaxes(showline=False, row=5, col=1)

    # Set the background to white
    fig.update_layout(plot_bgcolor='white')

    return fig

# Generate synthetic data for demonstration
true_labels = np.random.binomial(1, 0.5, 1000)
pred_probs = np.random.uniform(0, 1, 1000)

# Generate and show the modified faceted plot
faceted_fig_clean = runway_plot(true_labels, pred_probs)
faceted_fig_clean.show()
```

This will generate a plot like so:

![newplot](https://github.com/VectorInstitute/cyclops/assets/104445289/0b1c16ae-b25b-4eee-9cf2-deb6b79767ba)

**Comment by fcogidi** [comment]
@vaakesan-SMH Is your request that a `runway_plot` function (or something similar) should be added to `cyclops` (maybe as a utility function in the `evaluate` module)?

For your interest, I replaced `sklearn` metrics with `cyclops` metrics in the method you provided:

```python
import numpy as np
import plotly.graph_objects as go
import plotly.subplots as sp

from cyclops.evaluate.metrics.experimental.functional import binary_npv, binary_ppv
from cyclops.evaluate.metrics.functional import (
    binary_precision_recall_curve,
    binary_roc_curve,
)


def runway_plot(true_labels: np.ndarray, pred_probs: np.ndarray) -> go.Figure:
    """
    Plot diagnostic performance metrics with an additional histogram of predicted probabilities.
    The plot uses Plotly with a clean aesthetic. Gridlines are kept, but background color is removed.
    Y-axis ticks and labels are shown. The legend is added at the bottom.
    Tooltips show values with 3 decimal places. X-axis labels are only shown on the bottom subplot.
    The histogram's bin size is reduced and it has no borders.

    Args:
    - true_labels (np.ndarray): True binary class labels (0 or 1).
    - pred_probs (np.ndarray): Predicted probabilities for the positive class (1).

    Returns:
    - A Plotly figure containing the diagnostic performance plots and histogram.

    Example:
    ```
    # Generate synthetic data for demonstration
    true_labels = np.random.binomial(1, 0.5, 1000)
    pred_probs = np.random.uniform(0, 1, 1000)

    # Generate and show the modified faceted plot
    faceted_fig = plot_diagnostic_performance_with_histogram(true_labels, pred_probs)
    faceted_fig.show()
    ```
    """

    # ROC curve components
    fpr, tpr, _ = binary_roc_curve(true_labels, pred_probs)
    # Precision-Recall curve components
    precision, recall, _ = binary_precision_recall_curve(true_labels, pred_probs)

    # Thresholds for PPV and NPV
    thresholds = np.linspace(0, 1, 100)
    ppv = np.zeros_like(thresholds)
    npv = np.zeros_like(thresholds)

    # Calculate PPV and NPV for each threshold
    for i, threshold in enumerate(thresholds):
        # Calculate PPV and NPV
        ppv[i] = binary_ppv(true_labels, pred_probs, threshold=threshold)
        npv[i] = binary_npv(true_labels, pred_probs, threshold=threshold)

    # Define hover template to show three decimal places
    hover_template = "Threshold: %{x:.3f}<br>Metric Value: %{y:.3f}<extra></extra>"

    # Create a subplot for each metric
    fig = sp.make_subplots(rows=5, cols=1, shared_xaxes=True, vertical_spacing=0.02)

    # Sensitivity plot (True Positive Rate)
    fig.add_trace(
        go.Scatter(
            x=thresholds,
            y=tpr,
            mode="lines",
            name="Sensitivity",
            hovertemplate=hover_template,
        ),
        row=1,
        col=1,
    )

    # Specificity plot (1 - False Positive Rate)
    fig.add_trace(
        go.Scatter(
            x=thresholds,
            y=1 - fpr,
            mode="lines",
            name="1 - Specificity",
            hovertemplate=hover_template,
        ),
        row=2,
        col=1,
    )

    # PPV plot (Positive Predictive Value)
    fig.add_trace(
        go.Scatter(
            x=thresholds, y=ppv, mode="lines", name="PPV", hovertemplate=hover_template
        ),
        row=3,
        col=1,
    )

    # NPV plot (Negative Predictive Value)
    fig.add_trace(
        go.Scatter(
            x=thresholds, y=npv, mode="lines", name="NPV", hovertemplate=hover_template
        ),
        row=4,
        col=1,
    )

    # Add histogram of predicted probabilities
    fig.add_trace(
        go.Histogram(x=pred_probs, nbinsx=80, name="Predicted Probabilities"),
        row=5,
        col=1,
    )

    # Update layout
    fig.update_layout(
        height=1000,
        width=700,
        title_text="Diagnostic Performance Metrics by Thresholds",
        legend=dict(orientation="h", yanchor="bottom", y=-0.2, xanchor="center", x=0.5),
    )

    # Remove subplot titles
    for i in fig["layout"]["annotations"]:
        i["text"] = ""

    # Remove the plot background color, keep gridlines, show y-axis ticks and labels
    fig.update_xaxes(showgrid=True)
    fig.update_yaxes(showgrid=True, showticklabels=True)

    # Only show the x-axis line and labels on the bottommost plot
    fig.update_xaxes(showline=True, linewidth=1, linecolor="black", mirror=True)
    fig.update_xaxes(showticklabels=True, row=4, col=1)
    fig.update_yaxes(showline=True, linewidth=1, linecolor="black", mirror=True)

    fig.update_xaxes(showline=False, row=5, col=1, showticklabels=False)
    fig.update_yaxes(showline=False, row=5, col=1)

    # Set the background to white
    fig.update_layout(plot_bgcolor="white")

    return fig


# Generate synthetic data for demonstration
true_labels = np.random.binomial(1, 0.5, 1000)
pred_probs = np.random.uniform(0, 1, 1000)

# Generate and show the modified faceted plot
faceted_fig_clean = runway_plot(true_labels, pred_probs)
faceted_fig_clean.show()
```

**Comment by vaakesan-SMH** [comment]
Thanks @fcogidi. From my discussions with @amrit110, I believe this would be added as a method to the `ClassificationPlotter` class. Although perhaps it is better suited as a utility function in `evaluate` module. I am happy to defer to you and Amrit for the best option.

**Comment by fcogidi** [comment]
I think it's better to add it to the `ClassificationPlotter` class.

Would you like to add it yourself? That way you'll show up as a contributor.

**Comment by vaakesan-SMH** [comment]
Happy to submit a merge request. I will assign myself to this. Thanks Franklin.
